File: annotation-processor/src/main/java/me/bristermitten/mittenlib/annotations/compile/deserializer/GenericTypeDeserializerGenerator.java
package me.bristermitten.mittenlib.annotations.compile.deserializer;

import com.google.gson.reflect.TypeToken;
import com.google.inject.Inject;
import com.palantir.javapoet.CodeBlock;
import com.palantir.javapoet.MethodSpec;
import com.palantir.javapoet.TypeName;
import io.toolisticon.aptk.tools.TypeMirrorWrapper;
import io.toolisticon.aptk.tools.corematcher.AptkCoreMatchers;
import io.toolisticon.aptk.tools.wrapper.ElementWrapper;
import io.toolisticon.aptk.tools.wrapper.TypeElementWrapper;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import javax.lang.model.element.TypeElement;
import javax.lang.model.type.TypeMirror;
import me.bristermitten.mittenlib.annotations.ast.CustomDeserializerInfo;
import me.bristermitten.mittenlib.annotations.ast.Property;
import me.bristermitten.mittenlib.annotations.compile.ConfigurationClassNameGenerator;
import me.bristermitten.mittenlib.annotations.parser.CustomDeserializers;
import me.bristermitten.mittenlib.annotations.util.TypesUtil;
import me.bristermitten.mittenlib.config.CollectionsUtils;
import me.bristermitten.mittenlib.config.DeserializationFunction;
import me.bristermitten.mittenlib.config.tree.DataTree;
import me.bristermitten.mittenlib.config.tree.DataTreeTransforms;
import me.bristermitten.mittenlib.util.Result;
import me.bristermitten.mittenlib.util.Strings;

/**
 * Generates deserialization code for generic collection types (specifically {@link List} and {@link
 * Map}) where the element or value types are custom configuration types.
 */
public class GenericTypeDeserializerGenerator {

    private final TypesUtil typesUtil;
    private final ConfigurationClassNameGenerator configurationClassNameGenerator;
    private final CustomDeserializers customDeserializers;
    private final NonGenericTypeDeserializerGenerator nonGenericTypeDeserializerGenerator;

    @Inject
    GenericTypeDeserializerGenerator(
            TypesUtil typesUtil,
            ConfigurationClassNameGenerator configurationClassNameGenerator,
            CustomDeserializers customDeserializers,
            NonGenericTypeDeserializerGenerator nonGenericTypeDeserializerGenerator) {
        this.typesUtil = typesUtil;
        this.configurationClassNameGenerator = configurationClassNameGenerator;
        this.customDeserializers = customDeserializers;
        this.nonGenericTypeDeserializerGenerator = nonGenericTypeDeserializerGenerator;
    }

    /**
     * Creates a reference to a deserialization function, either as a method reference (for static) or
     * a field reference (for non-static, injected).
     *
     * <p>Generates:
     *
     * <pre>
     *     MyDeserializer::deserialize
     * </pre>
     *
     * <p>or
     *
     * <pre>
     *     this.myDeserializer
     * </pre>
     *
     * @param info the custom deserializer metadata, used to determine if the generated code should be
     *     static or not
     * @return a code block referencing the deserializer
     */
    private CodeBlock getDeserializationFunctionReference(CustomDeserializerInfo info) {
        if (info.isStatic()) {
            return CodeBlock.of("$T::deserialize", info.deserializerClass());
        }
        String fieldName =
                Strings.uncapitalize(info.deserializerClass().getSimpleName().toString());
        return CodeBlock.of("this.$L", fieldName);
    }

    /**
     * Generates and appends deserialization logic for generic collection properties.
     *
     * <p>For a {@link List}, generates:
     *
     * <pre>
     *     return CollectionsUtils.deserializeList(itemsFromMap, context, ctx0 -> ...);
     * </pre>
     *
     * @param builder the method spec builder
     * @param property the property being processed, used to generate the variable name (e.g. {@code
     *     itemsFromMap})
     * @param wrappedElementType the wrapped property type mirror
     * @param elementType the wrapped property type element
     * @return an optional method spec if handled successfully
     */
    public Optional<MethodSpec> handleGenericType(
            MethodSpec.Builder builder,
            TypeElement dtoType,
            Property property,
            TypeMirrorWrapper wrappedElementType,
            TypeElementWrapper elementType) {

        String canonicalName = wrappedElementType.erasure().getQualifiedName();
        ElementWrapper.wrap(property.source().element())
                .validate()
                .asError()
                .check($ -> AptkCoreMatchers.BY_RAW_TYPE
                        .getValidator()
                        .hasOneOf(elementType.unwrap(), List.class, Set.class, Map.class, Optional.class))
                .validateAndIssueMessages();

        final String fromMapName = property.name() + "FromMap";

        if (canonicalName.equals(List.class.getName())) {
            return handleListType(builder, dtoType, property, wrappedElementType, fromMapName);
        } else if (canonicalName.equals(Set.class.getName())) {
            return handleSetType(builder, dtoType, property, wrappedElementType, fromMapName);
        } else if (canonicalName.equals(Map.class.getName())) {
            return handleMapType(builder, dtoType, property, wrappedElementType, fromMapName);
        } else if (canonicalName.equals(Optional.class.getName())) {
            return handleOptionalType(builder, dtoType, property, wrappedElementType, fromMapName);
        } else {
            throw new IllegalStateException("Unexpected generic type: " + canonicalName);
        }
    }

    /**
     * Recursively generates a {@link CodeBlock} representing a {@link
     * DeserializationFunction} for the given type. Maps collections
     * recursively and falls back to object mapper mapping for basic types.
     *
     * <p>Generates lambdas like:
     *
     * <pre>
     *     ctx0 -> CollectionsUtils.deserializeList(ctx0.getData(), ctx0, ctx1 -> ...)
     * </pre>
     *
     * @param type the type to generate a deserialization function for
     * @param depth the current nesting depth, used to generate unique context variable names (e.g.
     *     {@code ctx0}, {@code ctx1})
     * @return a {@link CodeBlock} lambda expression {@code ctx -> ...}
     */
    private CodeBlock getDeserializationFunction(TypeElement dtoType, Property property, TypeMirror type, int depth) {
        TypeMirrorWrapper wrapped = TypeMirrorWrapper.wrap(type);

        // Custom Deserializer
        Optional<CustomDeserializerInfo> customDeserializerOptional = customDeserializers.getCustomInfo(type);
        if (customDeserializerOptional.isPresent()) {
            return getDeserializationFunctionReference(customDeserializerOptional.get());
        }

        // Config type
        if (typesUtil.isConfigType(type)) {
            String loaderField = configurationClassNameGenerator.getDeserializerProviderFieldName(type);
            return CodeBlock.of("this.$L.get()", loaderField);
        }

        // 3. Generic collections (List, Map)
        if (wrapped.hasTypeArguments()) {
            String canonicalName = wrapped.erasure().getQualifiedName();
            String ctxVar = "ctx" + depth;
            if (canonicalName.equals(List.class.getName())) {
                TypeMirror elementType = wrapped.getTypeArguments().getFirst();
                CodeBlock innerFunction = getDeserializationFunction(dtoType, property, elementType, depth + 1);
                return CodeBlock.of(
                        "$L -> $T.deserializeList($L.getData(), $L, $L)",
                        ctxVar,
                        CollectionsUtils.class,
                        ctxVar,
                        ctxVar,
                        innerFunction);
            } else if (canonicalName.equals(Set.class.getName())) {
                TypeMirror elementType = wrapped.getTypeArguments().getFirst();
                CodeBlock innerFunction = getDeserializationFunction(dtoType, property, elementType, depth + 1);
                return CodeBlock.of(
                        "$L -> $T.deserializeSet($L.getData(), $L, $L)",
                        ctxVar,
                        CollectionsUtils.class,
                        ctxVar,
                        ctxVar,
                        innerFunction);
            } else if (canonicalName.equals(Map.class.getName())) {
                var arguments = wrapped.getTypeArguments();
                TypeMirror keyType = arguments.get(0);
                TypeMirror valueType = arguments.get(1);
                CodeBlock innerFunction = getDeserializationFunction(dtoType, property, valueType, depth + 1);
                return CodeBlock.of(
                        "$L -> $T.deserializeMap($T.class, $L.getData(), $L, $L)",
                        ctxVar,
                        CollectionsUtils.class,
                        typesUtil.getSafeType(keyType),
                        ctxVar,
                        ctxVar,
                        innerFunction);
            } else if (canonicalName.equals(Optional.class.getName())) {
                TypeMirror elementType = wrapped.getTypeArguments().getFirst();
                CodeBlock innerFunction = getDeserializationFunction(dtoType, property, elementType, depth + 1);
                return CodeBlock.builder()
                        .add("($L) -> {\n", ctxVar)
                        .indent()
                        .beginControlFlow(
                                "if ($1L.getData() == null || $1L.getData() instanceof $2T)",
                                ctxVar,
                                DataTree.DataTreeNull.class)
                        .addStatement("return $T.ok($T.empty())", Result.class, Optional.class)
                        .endControlFlow()
                        .addStatement(
                                "return (($T) $L).apply($L).map($T::ofNullable)",
                                DeserializationFunction.class,
                                innerFunction,
                                ctxVar,
                                Optional.class)
                        .unindent()
                        .add("}")
                        .build();
            }
        }

        // 4. Basic fallback using ObjectMapper mapping
        String ctxVar = "ctx" + depth;
        MethodSpec.Builder innerMethod = MethodSpec.methodBuilder("temp");
        TypeName safeType = configurationClassNameGenerator.publicPropertyClassName(typesUtil.getBoxedType(type));
        boolean handled = nonGenericTypeDeserializerGenerator.handleNonGenericType(
                innerMethod, property, dtoType, type, wrapped, ctxVar + ".getData()", safeType);

        if (handled) {
            CodeBlock code = innerMethod.build().code();
            return CodeBlock.builder()
                    .add("$L -> {\n", ctxVar)
                    .add(code)
                    .add("}")
                    .build();
        }

        return CodeBlock.of(
                "$L -> $L.getMapper().map($L.getData(), new $T<$T>(){})",
                ctxVar,
                ctxVar,
                ctxVar,
                TypeToken.class,
                typesUtil.getBoxedType(type));
    }

    /**
     * Generates deserialization logic for a {@link List} property.
     *
     * <p>Generates:
     *
     * <pre>
     *     return CollectionsUtils.deserializeList(fromMap, context, ctx0 -> ...);
     * </pre>
     *
     * @param builder the method spec builder
     * @param wrappedElementType the wrapped property type mirror
     * @param fromMapName the name of the variable containing the raw data (e.g. {@code fromMap})
     * @return an optional method spec
     */
    private Optional<MethodSpec> handleListType(
            MethodSpec.Builder builder,
            TypeElement dtoType,
            Property property,
            TypeMirrorWrapper wrappedElementType,
            String fromMapName) {
        var listType = wrappedElementType.getTypeArguments().getFirst();
        CodeBlock deserializationFunction = getDeserializationFunction(dtoType, property, listType, 0);
        builder.addCode(
                "return $T.deserializeList($L, context, $L);\n",
                CollectionsUtils.class,
                fromMapName,
                deserializationFunction);
        return Optional.of(builder.build());
    }

    /**
     * Generates deserialization logic for a {@link Map} property.
     *
     * <p>Generates:
     *
     * <pre>
     *     return CollectionsUtils.deserializeMap(KeyType.class, fromMap, context, ctx0 -> ...);
     * </pre>
     *
     * @param builder the method spec builder
     * @param wrappedElementType the wrapped property type mirror
     * @param fromMapName the name of the variable containing the raw data (e.g. {@code fromMap})
     * @return an optional method spec
     */
    private Optional<MethodSpec> handleMapType(
            MethodSpec.Builder builder,
            TypeElement dtoType,
            Property property,
            TypeMirrorWrapper wrappedElementType,
            String fromMapName) {
        var arguments = wrappedElementType.getTypeArguments();
        var keyType = arguments.get(0);
        var valueType = arguments.get(1);

        CodeBlock deserializationFunction = getDeserializationFunction(dtoType, property, valueType, 0);
        builder.addCode(
                "return $T.deserializeMap($T.class, $L, context, $L);\n",
                CollectionsUtils.class,
                typesUtil.getSafeType(keyType),
                fromMapName,
                deserializationFunction);
        return Optional.of(builder.build());
    }

    private Optional<MethodSpec> handleSetType(
            MethodSpec.Builder builder,
            TypeElement dtoType,
            Property property,
            TypeMirrorWrapper wrappedElementType,
            String fromMapName) {
        var setType = wrappedElementType.getTypeArguments().getFirst();
        CodeBlock deserializationFunction = getDeserializationFunction(dtoType, property, setType, 0);
        builder.addCode(
                "return $T.deserializeSet($L, context, $L);\n",
                CollectionsUtils.class,
                fromMapName,
                deserializationFunction);
        return Optional.of(builder.build());
    }

    private Optional<MethodSpec> handleOptionalType(
            MethodSpec.Builder builder,
            TypeElement dtoType,
            Property property,
            TypeMirrorWrapper wrappedElementType,
            String fromMapName) {
        var optionalType = wrappedElementType.getTypeArguments().getFirst();
        CodeBlock deserializationFunction = getDeserializationFunction(dtoType, property, optionalType, 0);

        if (property.settings().hasDefaultValue()) {
            builder.beginControlFlow("if ($L instanceof $T)", fromMapName, Optional.class);
            builder.addStatement("return $T.ok(($T) $L)", Result.class, Optional.class, fromMapName);
            builder.endControlFlow();
        }

        builder.beginControlFlow("if ($L instanceof $T)", fromMapName, DataTree.DataTreeNull.class);
        builder.addStatement("return $T.ok($T.empty())", Result.class, Optional.class);
        builder.endControlFlow();

        String treeVar = property.name() + "Tree";
        if (property.settings().hasDefaultValue()) {
            builder.addStatement(
                    "$T $L = $T.loadFrom($L)", DataTree.class, treeVar, DataTreeTransforms.class, fromMapName);
        } else {
            builder.addStatement("$T $L = $L", DataTree.class, treeVar, fromMapName);
        }

        builder.addCode(
                "$T<$T> innerResult = $L.apply(context.withData($L));\n",
                Result.class,
                typesUtil.getBoxedType(optionalType),
                deserializationFunction,
                treeVar);
        builder.addStatement("return innerResult.map($T::ofNullable)", Optional.class);
        return Optional.of(builder.build());
    }
}

File: annotation-processor/src/main/scala/me/bristermitten/mittenlib/annotations/compile/deserializer/GenericTypeDeserializerGenerator.scala
package me.bristermitten.mittenlib.annotations.compile.deserializer

import com.google.gson.reflect.TypeToken
import com.google.inject.Inject
import com.palantir.javapoet.{
  ClassName,
  CodeBlock,
  MethodSpec,
  TypeName,
  ParameterizedTypeName
}
import io.toolisticon.aptk.tools.TypeMirrorWrapper
import io.toolisticon.aptk.tools.corematcher.AptkCoreMatchers
import io.toolisticon.aptk.tools.wrapper.{ElementWrapper, TypeElementWrapper}
import java.util.Optional
import javax.lang.model.element.{Modifier, TypeElement}
import javax.lang.model.`type`.TypeMirror
import _root_.me.bristermitten.mittenlib.annotations.ast.{
  AbstractConfigStructure,
  ConfigTypeSource,
  CustomDeserializerInfo,
  Property
}
import _root_.me.bristermitten.mittenlib.annotations.compile.ConfigurationClassNameGenerator
import _root_.me.bristermitten.mittenlib.annotations.parser.CustomDeserializers
import _root_.me.bristermitten.mittenlib.annotations.util.TypesUtil
import _root_.me.bristermitten.mittenlib.config.CollectionsUtils
import _root_.me.bristermitten.mittenlib.config.exception.ConfigLoadingErrors
import _root_.me.bristermitten.mittenlib.util.Strings

import _root_.me.bristermitten.mittenlib.codegen.dsl.*
import _root_.me.bristermitten.mittenlib.codegen.dsl.BlockBuilder.*
import _root_.me.bristermitten.mittenlib.codegen.dsl.ResultExpr
import _root_.me.bristermitten.mittenlib.codegen.dsl.given
import _root_.me.bristermitten.mittenlib.config.DeserializationContext

class GenericTypeDeserializerGenerator @Inject() (
    private val typesUtil: TypesUtil,
    private val configurationClassNameGenerator: ConfigurationClassNameGenerator,
    private val customDeserializers: CustomDeserializers,
    private val nonGenericTypeDeserializerGenerator: NonGenericTypeDeserializerGenerator
):

  private def getDeserializationFunction(
      dtoType: TypeElement,
      property: Property,
      tpe: TypeMirror,
      depth: Int
  ): Expr =
    val wrapped = TypeMirrorWrapper.wrap(tpe)
    val customDeserializerOptional = customDeserializers.getCustomInfo(tpe)
    if (customDeserializerOptional.isPresent) {
      val info = customDeserializerOptional.get()
      if (info.isStatic) {
        val ctxVar = Var("ctx", Types.DeserializationContext)
        Expr.lambdaExpr(
          Expr.staticCall(
            TypeRef.of(ClassName.get(info.deserializerClass)),
            "deserialize",
            ctxVar
          ),
          ctxVar
        )
      } else {
        val fieldName =
          Strings.uncapitalize(info.deserializerClass.getSimpleName.toString)
        Expr.This.field(fieldName)
      }
    } else if (typesUtil.isConfigType(tpe)) {
      val loaderField =
        configurationClassNameGenerator.getDeserializerProviderFieldName(tpe)
      Expr.This.field(loaderField).call("get")
    } else if (wrapped.hasTypeArguments && !typesUtil.isNewtype(tpe)) {
      val canonicalName = wrapped.erasure().getQualifiedName()
      val ctxVar = Var(s"ctx$depth", Types.DeserializationContext)
      if (canonicalName == classOf[java.util.List[?]].getName) {
        val elementType = wrapped.getTypeArguments.get(0)
        val innerFunction =
          getDeserializationFunction(dtoType, property, elementType, depth + 1)
        Expr.lambdaExpr(
          Expr.staticCall(
            TypeRef.of(classOf[CollectionsUtils]),
            "deserializeList",
            ctxVar.call("getData"),
            ctxVar,
            innerFunction
          ),
          ctxVar
        )
      } else if (canonicalName == classOf[java.util.Set[?]].getName) {
        val elementType = wrapped.getTypeArguments.get(0)
        val innerFunction =
          getDeserializationFunction(dtoType, property, elementType, depth + 1)
        Expr.lambdaExpr(
          Expr.staticCall(
            TypeRef.of(classOf[CollectionsUtils]),
            "deserializeSet",
            ctxVar.call("getData"),
            ctxVar,
            innerFunction
          ),
          ctxVar
        )
      } else if (canonicalName == classOf[java.util.Map[?, ?]].getName) {
        val keyType = wrapped.getTypeArguments.get(0)
        val valueType = wrapped.getTypeArguments.get(1)
        val innerFunction =
          getDeserializationFunction(dtoType, property, valueType, depth + 1)
        Expr.lambdaExpr(
          Expr.staticCall(
            TypeRef.of(classOf[CollectionsUtils]),
            "deserializeMap",
            Expr.staticField(
              TypeRef.of(TypeName.get(typesUtil.getSafeType(keyType))),
              "class"
            ),
            ctxVar.call("getData"),
            ctxVar,
            innerFunction
          ),
          ctxVar
        )
      } else {
        throw new IllegalStateException(
          "Unexpected nested generic type: " + canonicalName
        )
      }
    } else {
      val ctxVar = Var(s"ctx$depth", Types.DeserializationContext)
      val safeType = configurationClassNameGenerator.publicPropertyClassName(
        typesUtil.getBoxedType(tpe)
      )
      val lambdaBody = BlockBuilder.build {
        nonGenericTypeDeserializerGenerator.handleNonGenericType(
          property,
          dtoType,
          tpe,
          wrapped,
          ctxVar.call("getData"),
          (~ctxVar).as[DeserializationContext],
          safeType
        )
      }
      Expr.Lambda(List(ctxVar), lambdaBody)
    }

  def generateDeserializeMethod(
      propertyAST: AbstractConfigStructure,
      property: Property,
      dtoType: TypeElement,
      elementType: TypeMirror,
      wrappedElementType: TypeMirrorWrapper,
      elementTypeElement: TypeElementWrapper,
      safeType: TypeName,
      daoName: ClassName,
      fieldNameGenerator: _root_.me.bristermitten.mittenlib.annotations.compile.FieldNameGenerator,
      methodNames: _root_.me.bristermitten.mittenlib.annotations.compile.MethodNames
  ): MethodSpec =
    ElementWrapper
      .wrap(property.source().element())
      .validate()
      .asError()
      .check(el =>
        AptkCoreMatchers.BY_RAW_TYPE
          .getValidator()
          .hasOneOf(
            elementTypeElement.unwrap(),
            classOf[java.util.List[?]],
            classOf[java.util.Set[?]],
            classOf[java.util.Map[?, ?]]
          )
      )
      .validateAndIssueMessages()

    val elementResultType = TypeRef.of(safeType)
    val context = StagedExpr
      .param[DeserializationContext](Types.DeserializationContext, "context")

    val hasDefault = property.settings().hasDefaultValue()
    val dao = if (hasDefault && daoName != null) {
      Some(StagedExpr.param[Any](TypeRef.of(daoName), "dao"))
    } else {
      None
    }

    val params = List(context.asVar) ++ dao.map(_.asVar)

    val methodDecl = MethodDecl.build(
      name = methodNames.getDeserializeMethodName(property),
      returnType = Types.Result(elementResultType),
      parameters = params,
      modifiers = List(Modifier.PRIVATE)
    ) {
      // 1. Initial statements and Null checks
      val fromMap =
        _root_.me.bristermitten.mittenlib.annotations.compile.GeneratorUtil
          .declareAndCheckFromMap(
            propertyAST,
            property,
            dtoType,
            elementType,
            context,
            hasDefault,
            dao,
            fieldNameGenerator
          )

      // 3. Generic handling
      val canonicalName = wrappedElementType.erasure().getQualifiedName()

      if (canonicalName == classOf[java.util.List[?]].getName) {
        val listType = wrappedElementType.getTypeArguments().get(0)
        val deserializationFunction =
          getDeserializationFunction(dtoType, property, listType, 0)
        return_(
          Expr.staticCall(
            TypeRef.of(classOf[CollectionsUtils]),
            "deserializeList",
            fromMap,
            context,
            deserializationFunction
          )
        )
      } else if (canonicalName == classOf[java.util.Set[?]].getName) {
        val setType = wrappedElementType.getTypeArguments().get(0)
        val deserializationFunction =
          getDeserializationFunction(dtoType, property, setType, 0)
        return_(
          Expr.staticCall(
            TypeRef.of(classOf[CollectionsUtils]),
            "deserializeSet",
            fromMap,
            context,
            deserializationFunction
          )
        )
      } else if (canonicalName == classOf[java.util.Map[?, ?]].getName) {
        val keyType = wrappedElementType.getTypeArguments().get(0)
        val valueType = wrappedElementType.getTypeArguments().get(1)
        val deserializationFunction =
          getDeserializationFunction(dtoType, property, valueType, 0)
        return_(
          Expr.staticCall(
            TypeRef.of(classOf[CollectionsUtils]),
            "deserializeMap",
            Expr.staticField(
              TypeRef.of(TypeName.get(typesUtil.getSafeType(keyType))),
              "class"
            ),
            fromMap,
            context,
            deserializationFunction
          )
        )
      } else {
        throw new IllegalStateException(
          "Unexpected generic type: " + canonicalName
        )
      }
    }

    CodeBlockRenderer.renderMethod(methodDecl)
